src/typechecker.erl

```diff
@@ -3340,17 +3371,17 @@ infer_clause(Env, {clause, _, Args, Guards, Block}) ->
                                 end, GuardConj)
               end, Guards),
     {Ty, VB, Cs} = type_check_block(EnvNew, Block),
-    {Ty, union_var_binds(VB, VEnv, Env), Cs}.
+    {Ty, union_var_binds(VB, EnvNew, EnvNew), Cs}.
 
 
 check_clauses_intersect(Env, Ty, Clauses) when not is_list(Ty) ->
     check_clauses_fun(Env, Ty, Clauses);
-check_clauses_intersect(_Env, [], _Clauses) ->
-    {#{}, constraints:empty()};
+check_clauses_intersect(Env, [], _Clauses) ->
+    {Env, constraints:empty()};
 check_clauses_intersect(Env, [Ty|Tys], Clauses) ->
-    {VarBinds1, Cs1} = check_clauses_fun(Env, Ty, Clauses),
-    {VarBinds2, Cs2} = check_clauses_intersect(Env, Tys, Clauses),
-    {union_var_binds(VarBinds1, VarBinds2, Env), constraints:combine(Cs1, Cs2)}.
+    {Env1, Cs1} = check_clauses_fun(Env, Ty, Clauses),
+    {Env2, Cs2} = check_clauses_intersect(Env1, Tys, Clauses),
+    {union_var_binds(Env1, Env2, Env), constraints:combine(Cs1, Cs2)}.
```

Comment:
I'm fixing it, but the proposed test fails with an unrelated error (also on current master):

```
$ cat clauses_intersect.erl
-module(clauses_intersect).

-export([foo/1]).

-spec foo(bar) -> bar; (baz) -> baz.
foo(X = bar) -> X;
foo(X = baz) -> X.
$ ./bin/gradualizer --verbose clauses_intersect.erl
Checking module clauses_intersect
Spawning async task...
Checking function foo/1
Pushing {clauses_controls,true}
6:1: Checking clause :: bar
6:17: Checking that X :: bar
Task returned from function foo/1 with [{type_error,unreachable_clause,{7,1}}]
clauses_intersect.erl: The clause on line 7 at column 1 cannot be reached
```

Actually, now that I look at it I think `check_clauses_intersect` is buggy (or just incomplete? I think there's a comment somewhere in the code which suggests function intersections are not fully supported yet). Take an example like this:

```
-spec foo(bar) -> bar;
      foo(list()) -> baz | qux.
foo(bar) -> bar;
foo([]) -> baz;
foo([a]) -> qux;
foo([_, _ | T]) -> foo(T).
```

The question is: after how many fun clauses should we switch from checking against `foo(bar) -> bar` to `foo(list()) -> baz | qux`? Well, once our fun clauses exhaust `bar`. This is not done currently, hence the error returned when checking your example. We should probably do something more like this:

```erlang
try
    {Env1, Cs1} = check_clauses_fun(Env, Ty, Clauses),
catch
    throw:{type_error,unreachable_clause,...} ->
        %% Now try to type check the fun clauses against the next spec clause.
end
```

All in all, though, I think this is outside the scope of this PR.

Reply:
Agree that it's outside the scope of this PR.

We shouldn't switch, but choose spec depending on param pattern somehow. It's a function intersection type.

According to Erlang docs, the clauses are not allowed to overlap (though there's no check for that either) but this implies that a spec like this is illegal, because [] is in both domains: `-spec f([a]) -> a; ([b]) -> b.`

Reply:
For the record, this is related to #97, but not exactly the same thing. I've also created #410 to track this.